asllib/Interpreter.ml

```diff
@@ -251,86 +251,90 @@ module Make (B : Backend.S) (C : Config) = struct
-            let* vec = eval_expr env e' in
-            B.get_i i vec |: Rule.GetRecordField
+           let i = record_index_of_field e x li ty in
+           let* vec = eval_expr env e' in
+           B.get_i i vec |: Rule.GetRecordField
         | _ -> fatal_from e @@ Error.BadField (x, ty))
-    | E_GetFields (_, [], _) ->
-        V_BitVector (Bitvector.of_string "") |> B.v_of_parsed_v |> return
-    | E_GetFields (e', [ field ], ta) -> (
+      | E_GetFields (_, [], _) ->
+         V_BitVector (Bitvector.of_string "") |> B.v_of_parsed_v |> return
+      | E_GetFields (e', [ field ], ta) -> (
         let ty = type_of_ta e ta in
         match ty.desc with
         | T_Bits (_, fields) -> (
-            match List.assoc_opt field fields with
-            | Some slices ->
-                E_Slice (e', slices)
-                |> ASTUtils.add_pos_from e |> eval_expr env |: Rule.GetBitField
-            | None -> fatal_from e @@ Error.BadField (field, ty))
+          match List.assoc_opt field fields with
+          | Some slices ->
+             E_Slice (e', slices)
+             |> ASTUtils.add_pos_from e |> eval_expr env |: Rule.GetBitField
+          | None -> fatal_from e @@ Error.BadField (field, ty))
         | _ -> fatal_from e @@ Error.BadField (field, ty))
-    | E_GetFields (e', xs, ta) -> (
+      | E_GetFields (e', xs, ta) -> (
         let ty = type_of_ta e ta in
         match ty.desc with
         | T_Bits (_, fields) ->
-            let one (x : string) =
-              match List.assoc_opt x fields with
-              | None -> fatal_from e @@ Error.BadField (x, ty)
-              | Some slices ->
-                  E_Slice (e', slices)
-                  |> ASTUtils.add_pos_from e |> eval_expr env
-            in
-            prod_map one xs >>= B.concat_bitvectors |: Rule.GetBitFields
+           let one (x : string) =
+             match List.assoc_opt x fields with
+             | None -> fatal_from e @@ Error.BadField (x, ty)
+             | Some slices ->
+                E_Slice (e', slices)
+                |> ASTUtils.add_pos_from e |> eval_expr env
+           in
+           prod_map one xs >>= B.concat_bitvectors |: Rule.GetBitFields
         | _ -> fatal_from e @@ Error.BadField (List.hd xs, ty))
-    | E_Concat es ->
-        prod_map (eval_expr env) es >>= B.concat_bitvectors |: Rule.EConcat
-    | E_Tuple _ -> fatal_from e @@ Error.NotYetImplemented "tuple construction"
-    | E_Unknown ty -> base_value_of_type env ty
-    | E_Pattern (e, p) ->
-        let* v = eval_expr env e in
-        eval_pattern env e v p
+      | E_Concat es ->
+         prod_map (eval_expr env) es >>= B.concat_bitvectors |: Rule.EConcat
+      | E_Tuple _ -> fatal_from e @@ Error.NotYetImplemented "tuple construction"
+      | E_Unknown ty -> base_value_of_type env ty
+      | E_Pattern (e, p) ->
+         let* v = eval_expr env e in
+         eval_pattern env e v p in
+    if false then
```

Comment:
This adds a monadic bind at every expression evaluated, could we instead do something like:
```ocaml
  let _debug_eval_expr m =
    let* v = m in
    let () =
      Format.eprintf "@[<3>Eval@ @[%a@] ->@ %s@]@."
        PP.pp_expr e (B.debug_value v)
    in
    return v

  let rec eval_expr (env : env) (e : expr) : B.value B.m =
     (if false then _debug_eval_expr else Fun.id)
     @@
     match e.desc with
     ...
```

Reply:
Yes it would be great. Given the important changes of the interpreter in the concurrent PR #598, I'd rather revert this change for now.